### ISSUE TITLE
initdb: error: could not change permissions of directory "/home/postgres/pgdata/data": Operation not permitted

### ISSUE THREAD

**Issue by sawjam** [timescale/timescaledb-docker-ha#359]
using image: timescale/timescaledb-ha:pg15.2-ts2.9.3-patroni-dcs-failsafe-latest
windows 11 same issue re permissions

**Comment by sawjam** [comment]
Full Error:
2023-02-10 10:27:01 The files belonging to this database system will be owned by user "postgres".
2023-02-10 10:27:01 This user must also own the server process.
2023-02-10 10:27:01 
2023-02-10 10:27:01 The database cluster will be initialized with locale "C.UTF-8".
2023-02-10 10:27:01 The default database encoding has accordingly been set to "UTF8".
2023-02-10 10:27:01 The default text search configuration will be set to "english".
2023-02-10 10:27:01 
2023-02-10 10:27:01 Data page checksums are disabled.
2023-02-10 10:27:01 
2023-02-10 10:27:01 fixing permissions on existing directory /home/postgres/pgdata/data ... 2023-02-09T21:27:01.831977200Z initdb: error: could not change permissions of directory "/home/postgres/pgdata/data": Operation not permitted

**Comment by sawjam** [comment]
Note: On a Windows 11 system if you run the docker run command from within a wsl2 env (Ubuntu in my case) and use a linux directory in the wsl2 env (still accessible from windows explorer) then the container builds with no errors and remains up, dropped back to using timescale/timescaledb-ha:pg15-latest, but worked on previous timescale/timescaledb-ha:pg15.2-ts2.9.3-patroni-dcs-failsafe-latest image too.

**Comment by sawjam** [comment]
FYI
Windows 11 (fails on permissions)
docker run --name timescaledb -e PGDATA=/home/pgdata POSTGRES_PASSWORD=password -p 5432:5432 -v c:\postgres\data:/home/pgdata -d timescale/timescaledb-ha:pg15-latest

WSL2 (ubuntu venv on windows 11 - works fine)
docker run --name timescaledb -e POSTGRES_PASSWORD=password -p 5432:5432 -v /home/timescaledb:/home/postgres/pgdata/data -d timescale/timescaledb-ha:pg15-latest

**Comment by natebowang** [comment]
Doesn't work for me too, just pulled the latest image. Here is my Docker compose:
```
  store-o:
    image: timescale/timescaledb-ha:pg14-latest
    restart: always
    environment:
      - POSTGRES_USER=pgroot
      - POSTGRES_PASSWORD=pgroot
    volumes:
      - ./store/.docker/localData/:/home/postgres/pgdata/
```

**Comment by natebowang** [comment]
I also tried: 
    user: 0:0
    user: 1000:1000
All don't work.
Hosted on Archlinux

**Comment by graveland** [comment]
@natebowang - please try by using a different destination for the mount, `/home/postgres/pgdata` is owned by `postgres` and can't be changed since the container is running as an unprivileged user `postgres`. If you add `PGDATA: /anywhere/but/home/postgres/pgdata` and change the mount point, does that work for you?

**Comment by coutoPL** [comment]
I can confirm. The same issue is in my case. I try to use `timescale/timescaledb-ha:pg15.2-ts2.9.3-patroni-dcs-failsafe-p0`

**Comment by graveland** [comment]
Could you try with `timescaledev/timescaledb-ha:pg15-ts2.9-all`?

**Comment by sawjam** [comment]
Same problem using:timescaledev/timescaledb-ha:pg15-ts2.9-all

2023-02-21 16:20:45 fixing permissions on existing directory /home/other/postgres/pgdata/data ... 2023-02-21T03:21:12.144337200Z The files belonging to this database system will be owned by user "postgres".
2023-02-21 16:21:12 This user must also own the server process.

**Comment by MarSchra** [comment]
Hi there, 

same problem with pg15-latest, pg14-latest under wsl2. Also tried to mount different folders via  
-e PGDATA /home/postgres/test
-e PGDATA /home/postgres/
-e PGDATA /var/lib/postgresql/data

docker run --name timescaledb -e POSTGRES_PASSWORD=password -e PGUSER=postgres -e PGDATA=/var/lib/postgresql/data/ -p 127.0.0.1:5432:5432 -v ~/home/myuser/TimescaleDB/Tests:/var/lib/postgresql/data/ timescale/timescaledb-ha:pg15-latest 

All of these mount points gave the message "error: could not change permissions of directory ..."

**Comment by drumsta** [comment]
I can confirm that the same issue is present in my case. I tried to use `timescale/timescaledb-ha:pg15-latest`.

**Comment by codefaux** [comment]
Suffered the same bug. Got it to run.  TLDR; chmod your datadir to 1000.

Relevant bits:

```
mkdir /docker/timescaledb
chown 1000 /docker/timescaledb
docker run -d --name timescaledb -p 5432:5432 -v /docker/timescaledb:/home/postgres/pgdata/data timescale/timescaledb-ha:pg15-latest
```

Long version; Container's UID map in /etc/passwd uses 1000 as postgres. Host OS doesn't always do that -- in my case, 1000 is used for other things, so just chown by UID instead of by username.

Docker (or the container, or _whatever_) creates your target volume mount as a directory with root ownership, then it gets mounted into the container. The container process is running UID 1000, and tries to change ownership but can't because it's owned by root. Set the ownership to 1000 yourself and it won't complain.

Not sure if the docker image can be configured to create its mounted volumes as UID=1000 but that's the upstream fix.

**Comment by jankap** [comment]
This does not work for me unfortunately:

```
    timescaledb:
        image: timescale/timescaledb-ha:pg14-latest
        hostname: timescaledb
        volumes:
            - ./postgresql-data:/home/postgres/pgdata/data
        expose:
            - "5432"
```

with `chown 1000 ./postgresql-data` or `sudo chown` in WSL2.

I need to use a Docker volume to make it work.

**Comment by codefaux** [comment]
I'm not sure what the differences are here. I just did it from scratch in an empty dir etc etc and took some screenshots, maybe something is different?

docker-compose;
![image](https://user-images.githubusercontent.com/42219549/229205820-e8f66e99-6cba-4249-974c-93f86c32c392.png)


terminal logs;
![image](https://user-images.githubusercontent.com/42219549/229205679-02453cc3-3b61-4a26-b8ee-8b7e09a3df3e.png)



aftermath;
![image](https://user-images.githubusercontent.com/42219549/229206108-d07d93ea-5db2-4929-b1db-1e6a8517d8b4.png)

**Comment by orpheousff8** [comment]
> EDIT: This is real Linux not WSL, maybe that's the difference? What is your WSL setup? Just standard Docker-for-Windows? I'll replicate on my end as a test if you share.
> 
> I'm not sure what the differences are here. I just did it from scratch in an empty dir etc etc and took some screenshots, maybe something is different?
> 
> docker-compose; ![image](https://user-images.githubusercontent.com/42219549/229205820-e8f66e99-6cba-4249-974c-93f86c32c392.png)
> 
> terminal logs; ![image](https://user-images.githubusercontent.com/42219549/229205679-02453cc3-3b61-4a26-b8ee-8b7e09a3df3e.png)
> 
> aftermath; ![image](https://user-images.githubusercontent.com/42219549/229206108-d07d93ea-5db2-4929-b1db-1e6a8517d8b4.png)

Yes, if it's Docker on Linux it works if did `chown 1000 data`. How ever, on WSL2 it keeps posting this error:

```timescale-db-tracking  | Waiting for permissions on /home/postgres/pgdata/data (0:0 -> 1000:1000)........................................
timescale-db-tracking  | 2023-04-27 10:10:04.942 UTC [1] FATAL:  data directory "/home/postgres/pgdata/data" has wrong ownership
timescale-db-tracking  | 2023-04-27 10:10:04.942 UTC [1] HINT:  The server must be started by the user that owns the data directory.
timescale-db-tracking exited with code 1```

**Comment by orpheousff8** [comment]
> Okay, I know how to fix this on Windows now, I need to map volume to _/data_, **in stead of** _/var/lib/postgresql/data_ like I used to do with timescale/timescaledb:latest-pg13.

FYI, for your questions.
- Docker images of timescale DB I tested was timescale/timescaledb-ha:pg13.10-ts2.10.3-all
- WSL version 1.2.5.0 on Windows 10.0.22621.1555 (Win11)
- Which WSL distribution/version are you running? 
     I run Docker Desktop 4.18.0 on WSL
- Where is your data directory?
     I mounted "C:\Users\myUser\pgdata-ha" to "/home/postgres/pgdata/data"

Though my file system is NTFS, I run docker and set file permission / ownership like in UNIX from Git Bash.

I can run Docker image of timescale/timescaledb:latest-pg13, mySQL, mariaDB, Wordpress, Ruby, Debian on this system settings with no issues.

**Comment by codefaux** [comment]
Not at the computer and unable to test now/today but /data is not the correct mount and if I'm right, while the container _will_ run, /data won't populate and your database will be lost on recreation/updating of the container.

If you were mounting the wrong point (as you are now) it would silently not notice because it has no way to know you've mounted an extra, incorrect path (as it is now) because it didn't use that path for anything..as opposed to mounting the correct path which it can't work with properly for whatever reason and causing an error -- as like it was before.

You can and will accomplish the exact same success by not mounting any volumes and starting the container, and you'll lose your database just the same. If you can deal with the database contents randomly disappearing, proceed as you are now. The "where is your data directory" question is the one that's most likely going to lead to solving this.

**Comment by graveland** [comment]
You definitely want to mount a data volume somewhere other than `/home/postgres/pgdata/data`, but to get it to work, you need to also set the `PGDATA` environment variable to the place you mounted. By using a different location outside of `/home/postgres` the bits that allow non-UID-1000 users to work should be able to succeed.

(I haven't tested that in a while though)

**Comment by orpheousff8** [comment]
Fixed and be able to persist the data. Need to specify user as _**root**_, and change data directory (I used _/var/lib/postgresql/data_). I guess the default user _postgres_ don't have a permission to write Windows directory.

![image](https://user-images.githubusercontent.com/25906655/235263847-430e16d6-1ad4-471f-8dc1-c0591985ba9c.png)

**Comment by BeeStingBob** [comment]
was getting this error 
`"initdb: error: could not change permissions of directory "/home/postgres/pgdata/data": Operation not permitted fixing permissions on existing directory /home/postgres/pgdata/data ... The files belonging to this database system will be owned by user "postgres". This user must also own the server process"`

after reviewing this
https://github.com/timescale/timescaledb-docker/blob/main/bitnami/Dockerfile
and seeing the user being set to "USER 1001" line 110
added the --user 1011 (the uid of gitlab_user) to my docker run command  and it worked

`docker run -id --user 1011  --name timescaledb -v /home/gitlab_user/postgresdata:/home/postgres/pgdata/data --restart unless-stopped timescale/timescaledb-ha:pg14-latest`

**Comment by wenwengongren270** [comment]
You need to first setup PGDATA(in your local machine) right access. sudo chown -R dev:dev ~/PGDATA，and then run the docker start command will work. The correct user of PGDATA should not root if you run docker command in other user.

**Comment by FStefanni** [comment]
Hi,

I had the same problem with `timescale/timescaledb-ha:pg15.4-ts2.11.2`.
I have solved in this way:

1. Created a local dir: `mkdir -p /somepath/data`
2. Changed the permissions to all: `chmod -R a+rwx /somepath/data`
3. Then run the container with the following config:
    * PGDATA=/somepath-inside-container/data/pgdata
    * Volume: /somepath/data:/somepath-inside-container/data

The trick has been to create `data` with all permissions, and let the container create the `pgdata` subdirectory.
Hope this helps.

Regards

**Comment by cbadawi** [comment]
I'm running into this issue if i change the pg password to anything other than "password"

```
sudo docker run -d --name timescaledb -p 5432:5432 -e POSTGRES_PASSWORD=password -v /var/local/timescale/data:/home/postgres/pgdata/data timescale/timescaledb-ha:pg14-latest
```

changing PGDATA did not solve it for me

**Comment by ant0nm** [comment]
Hey guys, I'm running into the exact same permissions issue on `initdb`.
Image I'm using: `timescale/timescaledb-ha:pg15.2-ts2.10.2-latest`.
Is there any update on a workaround/fix for this?

**Comment by laythhamad** [comment]
I recently encountered a challenge with Docker deployment on Windows 11, specifically while attempting to mount an external directory on a Windows 11 host using WSL2 (Windows Subsystem for Linux 2) as the backend. This situation involves a dual-mounting process: firstly from the host OS to WSL2, and subsequently from WSL2 to Docker.

To effectively address this issue, I followed these steps:

Initiate WSL Command Line: Begin by opening the Start menu and launching the WSL command line interface.

Modify Directory Ownership in WSL: Change to the directory you intend to mount. For accessing directories within the Windows file system from WSL, navigate using the path /mnt/host/c/<your directory path>.

Update Ownership Permissions: Execute the command chown -R 1000 <your directory>/ to change the ownership of the directory, setting it to user ID 1000.

Finalize Permission Configuration in Docker: After completing these steps, Docker's initialization stage will be able to modify the permissions for the specified directory, ensuring proper access and functionality.

These steps successfully resolved the mounting issue in my Docker deployment on Windows 11 with WSL2 as the backend.
